porch/api/porchconfig/v1alpha1/types_functions.go

```diff
@@ -0,0 +1,107 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:path=functions,singular=function
+
+// Function represents a kpt function discovered in a repository
+// Function resources are created automatically by discovery in a registered Repository.
+// Function resource names will be computed as <Repository Name>:<Function Name>
+// to ensure uniqueness of names, and will follow formatting of
+// [DNS Subdomain Names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
+type Function struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FunctionSpec   `json:"spec,omitempty"`
+	Status FunctionStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// FunctionList
+type FunctionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Function `json:"items"`
+}
+
+type FunctionType string
+
+const (
+	FunctionTypeValidator FunctionType = "validator"
+	FunctionTypeMutator   FunctionType = "mutator"
+)
+
+// FunctionSpec defines the desired state of a Function
+type FunctionSpec struct {
+	// Image specifies the function image, such as 'gcr.io/kpt-fn/gatekeeper:v0.2'.
+	Image string `json:"image"`
+
+	// RepositoryRef references the repository in which the function is located.
+	RepositoryRef RepositoryRef `json:"repositoryRef"`
+
+	// FunctionType specifies the function types (mutator, validator or/and others).
+	FunctionTypes []FunctionType `json:"functionTypes,omitempty"`
+
+	FunctionConfigs []FunctionConfig `json:"functionConfigs,omitempty"`
+
+	// Keywords are used as filters to provide correlation in function discovery.
+	Keywords []string `json:"keywords,omitempty"`
+
+	// Description is a short description of the function.
+	Description string `json:"description"`
+
+	// `DocumentationUrl specifies the URL of comprehensive function documentation`
+	DocumentationUrl string `json:"documentationUrl,omitempty"`
+
+	// InputTypes specifies to which input KRM types the function applies. Specified as Group Version Kind.
+	// For example:
+	//
+	//    inputTypes:
+	//    - kind: RoleBinding
+	//      # If version is unspecified, applies to all versions
+	//      apiVersion: rbac.authorization.k8s.io
+	//    - kind: ClusterRoleBinding
+	//      apiVersion: rbac.authorization.k8s.io/v1
+	// InputTypes []metav1.TypeMeta
```

Comment:
Maybe something like this belongs in FunctionConfig type below?

Reply:
Ah - this file is the aggregated definition, just moved here: https://github.com/GoogleContainerTools/kpt/blob/921abb821a8663f5301d7e2f2d87846122dc7e30/porch/api/porch/v1alpha1/types_functions.go

I'm trying to not change it during the transition to CRDs (though I could)

Reply:
Oh, ok. Better in a separate PR if we think we want to change that. I am assuming these as CRDs is intended to allow discovery of functions, so in that case, understanding the CRDs that are inputs to those seems critical.